pymc_experimental/marginal_model.py

```diff
@@ -0,0 +1,440 @@
+
+def _collect_updates(rvs: Sequence[TensorVariable]) -> Dict[TensorVariable, TensorVariable]:
+    rng_updates = {}
+    for rv in rvs:
+        if isinstance(rv.owner.op, RandomVariable):
+            rng = rv.owner.inputs[0]
+            assert not hasattr(rng, "default_update")
+            rng_updates[rng] = rv.owner.outputs[0]
+        elif isinstance(rv.owner.op, SymbolicRandomVariable):
+            rng_updates.update(rv.owner.op.update(rv.owner))
+        else:
+            raise TypeError(f"Unknown RV type: {rv.owner.op}")
+    assert all(
+        isinstance(rng, (RandomStateSharedVariable, RandomGeneratorSharedVariable))
+        for rng in rng_updates.keys()
+    )
+    return rng_updates
+
+
+def replace_finite_discrete_marginal_subgraph(fgraph, rv_to_marginalize, all_rvs):
+    # TODO: This should eventually be integrated in a more general routine that can
+    #  identify other types of supported marginalization, of which finite discrete
+    #  RVs is just one
+
+    dependent_rvs = find_conditional_dependent_rvs(rv_to_marginalize, all_rvs)
+    if not dependent_rvs:
+        raise ValueError(f"No RVs depend on marginalized RV {rv_to_marginalize}")
+
+    ndim_supp = {rv.owner.op.ndim_supp for rv in dependent_rvs}
+    if max(ndim_supp) > 0:
+        raise NotImplementedError(
+            "Marginalization of withe dependent Multivariate RVs not implemented"
+        )
+
+    marginalized_rv_input_rvs = find_conditional_input_rvs([rv_to_marginalize], all_rvs)
+    dependent_rvs_input_rvs = [
+        rv
+        for rv in find_conditional_input_rvs(dependent_rvs, all_rvs)
+        if rv is not rv_to_marginalize
+    ]
+
+    # If the marginalized RV has batched dimensions, check that graph between
+    # marginalized RV and dependent RVs is composed strictly of Elemwise Operations.
+    # This implies (?) that the dimensions are completely independent and a logp graph
+    # can ultimately be generated that is proportional to the support domain and not
+    # to the variables dimensions
+    # We don't need to worry about this if the  RV is scalar.
+    if np.prod(constant_fold(tuple(rv_to_marginalize.shape))) > 1:
+        if not is_elemwise_subgraph(rv_to_marginalize, dependent_rvs_input_rvs, dependent_rvs):
+            raise NotImplementedError(
+                "The subgraph between a marginalized RV and its dependents includes non Elemwise operations. "
+                "This is currently not supported",
+            )
+
+    input_rvs = [*marginalized_rv_input_rvs, *dependent_rvs_input_rvs]
+    rvs_to_marginalize = [rv_to_marginalize, *dependent_rvs]
+
+    outputs = rvs_to_marginalize
+    # Clone replace inner RV rng inputs so that we can be sure of the update order
+    # replace_inputs = {rng: rng.type() for rng in updates_rvs_to_marginalize.keys()}
+    # Clone replace outter RV inputs, so that their shared RNGs don't make it into
+    # the inner graph of the marginalized RVs
+    # FIXME: This shouldn't be needed!
```

Comment:
Should this FIXME be addressed before merging?

Reply:
It's an Aesara thing, so definitely not in this PR